MaxText/microbenchmark.py

```diff
@@ -0,0 +1,184 @@
+
+  # Warmup
+  decode_state, _ = engine.generate(params, decode_state)
+  jax.block_until_ready(decode_state)
+  decode_state, _ = engine.generate(params, decode_state)
+  jax.block_until_ready(decode_state)
+
+  profile_name = "autoregress" if profile_name == "" else profile_name
+  time_in_s = ar_benchmark_loop(engine, decode_state, params, profile_name=profile_name, steps=steps)
+  seconds_per_step = time_in_s / config.steps
+  ar_average_ms = seconds_per_step*1000
+  total_throughput = jax.device_count() * config.per_device_batch_size / seconds_per_step
+
+  GB_per_step_per_device = (model_size + cache_size) / 1e9 / jax.device_count()
+  bw_per_device = GB_per_step_per_device/seconds_per_step
+  print(f"AutoRegressive results:\n"
+        f"\tAR step average time: {ar_average_ms:.2f}ms\n"
+        f"\tAR global batch size: {global_batch_size}\n"
+        f"\tAR throughput: {total_throughput:.2f} tokens/second\n"
+        f"\tAR memory bandwidth per device: {bw_per_device:.2f} GB/s")
+  return {"ar_step_in_ms": ar_average_ms, 
+          "ar_global_batch_size": global_batch_size, 
+          "ar_total_throughput_tokens_per_second": total_throughput,
+          "ar_device_bandwidth_GB_per_second": bw_per_device}
+
+
+def main(config):
+  engine = maxengine.MaxEngine(config)
+  params = engine.load_params()
+  prefill_lengths = [128, 256, 512, 1024]
+  num_steps = 10
+  text = config.prompt
+  metadata = engine.get_tokenizer()
+  vocab = token_utils.load_vocab(metadata.path, metadata.extra_ids)
+
+  decode_state = engine.init_decode_state()
+  delete_pytree(decode_state)
+  _, cache_size, _ = summarize_pytree_data(decode_state['cache'], name="Cache", log=True)
+  num_model_params, model_size, _ = summarize_pytree_data(params, name="Params", log=True)
+  results = {"config": {}, 
+             "sizes": {
+                "model_size_in_GB": model_size / 1e9,
+                "cache_size_in_GB": cache_size / 1e9,
+                "num_model_params_in_billions": num_model_params / 1e9,
+             }}
+  for k, v in dict(config.get_keys()).items():
+    results["config"][k] = str(v) if k == "dtype" else v
+
+  ar_results = ar_benchmark(config, engine, params, steps = num_steps, cache_size=cache_size, model_size=model_size)
+
+  for prefill_length in prefill_lengths:
+    tokens, true_length = token_utils.tokenize_and_pad(text, vocab, is_bos=True, prefill_lengths=[prefill_length])
+    prefill_results = prefill_benchmark(config, engine, params, tokens, true_length, num_steps, num_model_params)
+    prefill_results.update(ar_results)
+    results[prefill_length] = prefill_results
+
+  with open("benchmark_results.json", "w") as f:
+    json.dump(results, f, indent=2)
+
+
+if __name__ == "__main__":
+  jax.config.update('jax_default_prng_impl', 'unsafe_rbg')
+  os.environ["TF_CPP_MIN_LOG_LEVEL"] = "0"
+  os.environ["M_ICI_FSDP_PARALLELISM"] = "1"
+  os.environ["M_ICI_AUTOREGRESSIVE_PARALLELISM"] = "-1"
```

Comment:
pretty intense. I'd like you to hold off on this. In the future we might want different settings.